server/rpc/SessionRPC.java

```diff
@@ -97,11 +97,15 @@ private void setIdleTimeout() {
     }
 
     private void triggerIdleTimeout() {
+        if (!transactionRPCs.isEmpty()) {
+            keepAlive();
+            return;
+        }
         close();
         LOG.warn("Session with ID " + session.uuid() + " timed out due to inactivity");
     }
 
-    void keepAlive() {
+    synchronized void keepAlive() {
```

Comment:
Now that `triggerIdleTimeout` has a chance to call `keepAlive`, it's possible for `keepAlive` to be called concurrently by a (client-triggered) **session pulse** and a (server-triggered) **idle check**. This could lead to concurrent access of `setIdleTimeout` which in turn could lead to scheduling two tasks when we only need one.